lib/http-bridge/test/integration/Test/Integration/Scenario/CLI/Addresses.hs

```diff
@@ -10,41 +10,205 @@ module Test.Integration.Scenario.CLI.Addresses
+                      , "USED"
+                      , "UNUSED"
+                      , "-1000"
+                      , "44444444"
+                      , "*"
+                      ]
+
+        forM_ filters $ \fil -> it ("--state=" <> fil) $ \ctx -> do
+            walId <- emptyWallet' ctx
+            (Exit c, Stdout o, Stderr e)
+                <- listAddressesViaCLI ["--state", fil, walId]
+            e `shouldBe` "Unable to decode address state: it's neither\
+                \ \"used\" nor \"unused\"\n"
+            c `shouldBe` ExitFailure 1
+            o `shouldBe` ""
+
+    it "ADDRESS_LIST_03 - Generates new address pool gap" $ \ctx -> do
+        let initPoolGap = 10
+        wSrc <- fixtureWallet' ctx
+        wDest <- emptyWalletWith ctx ("Wallet", "cardano-wallet", initPoolGap)
+        let widDest = T.unpack $ wDest ^. walletId
+
+        -- make sure all addresses in address_pool_gap are 'Unused'
+        (Exit c, Stdout o, Stderr e) <- listAddressesViaCLI [widDest]
+        e `shouldBe` "Ok.\n"
+        c `shouldBe` ExitSuccess
+        j <- expectValidJSON (Proxy @[ApiAddress t]) o
+        length j `shouldBe` initPoolGap
+        forM_ [0..initPoolGap - 1] $ \addrNum -> do
+            expectCliListItemFieldEqual addrNum state Unused j
+
+        -- run 10 transactions to make all addresses `Used`
+        forM_ [0..initPoolGap - 1] $ \addrNum -> do
+            let dest = encodeAddress (Proxy @t) (getApiT $ fst $ (j !! addrNum) ^. #id)
+            let args = [wSrc, "--payment" , T.unpack $ "1@" <> dest]
+            (cTx, _, _) <- postTransactionViaCLI "cardano-wallet" args
+            cTx `shouldBe` ExitSuccess
+
+        -- make sure all transactions are in ledger
+        expectEventually' ctx balanceAvailable 10 wDest
+
+        -- verify new address_pool_gap has been created
+        (Exit c1, Stdout o1, Stderr e1) <- listAddressesViaCLI [widDest]
+        e1 `shouldBe` "Ok.\n"
+        c1 `shouldBe` ExitSuccess
+        j1 <- expectValidJSON (Proxy @[ApiAddress t]) o1
+        length j1 `shouldBe` 2*initPoolGap
+        forM_ [0..initPoolGap - 1] $ \addrNum -> do
+            expectCliListItemFieldEqual addrNum state Used j1
+        forM_ [initPoolGap..2*initPoolGap - 1] $ \addrNum -> do
+            expectCliListItemFieldEqual addrNum state Unused j1
+
+    describe "ADDRESS_LIST_04 - False wallet ids" $ do
+        forM_ falseWalletIds $ \(title, walId) -> it title $ \_ -> do
+            (Exit c, Stdout o, Stderr e) <- listAddressesViaCLI [walId]
+            o `shouldBe` ""
+
+            if (title == "40 chars hex") then
+                e `shouldBe` errMsg404NoWallet "1111111111111111111111111111111111111111\n"
+            else
+                e `shouldBe` "wallet id should be an hex-encoded string of\
+                    \ 40 characters\n"
+
+            if (title == "40 chars hex") then
+                c `shouldBe` ExitSuccess
+            else
+                c `shouldBe` ExitFailure 1
```

Comment:
Quite unrelated but this is odd? Do we return a success on 404?

Reply:
Yes, I think so too, but I was told that this is expected... ;) -> https://github.com/input-output-hk/cardano-wallet/pull/278#discussion_r285063171

Reply:
Then I've very poorly expressed myself. I meant that you were right and that this oddity is not correct. May you open a bug ticket to keep track of that please?

Reply:
Ah, ok. My bad for misunderstanding then. Will open a bug :+1: